### DISCUSSION TITLE
Is it possible to have a source json where you copy a subkey to a new json?

### DISCUSSION THREAD

**Discussion by jasonmnemonic** [danielaparker/jsoncons#312] [Q&A]
Hi Daniel.

Thanks for the great library and hard work. It is a cool library. ;-)

A question on json and maybe anyone can answer this (thanks): 

Say, I have a source json:
```
{
        "nameid":1,
        "name":"platypus",
        "postcode":1,
        "previous-postcode":
            {
                "postcode1":0,
                "postcode2": 5,
                "postcode3": 9
            }
}
```
When I make a new `jsoncons::json originalJsonDoc` using `json::parse` to parse the string above. Next, I create a new `jsoncons:json newJsonDoc` that is empty.

In my new json, I am interested to `newJsonDoc[ "previous-postcode" ] =  jsonRoot[ "previous-postcode" ].to_string();` but this returns me a string like `{"previous-postcode": "{ \"postcode1\":0, \"postcode2\":5, \"postcode2\":9 }"}`.

Is there a way to do something like `newJsonDoc[ "previous-postcode" ] =  jsonRoot[ "previous-postcode" ]` so that when I `dump-pretty`, I will get  `{"previous-postcode": {"postcode1":0, "postcode2":5, "postcode3":9 }}`.

I think there is a way but I am having trouble to locate it.

Thank you! ;-)

**Comment by danielaparker** [comment]
The assignment
```
newJsonDoc["previous-postcode"] = jsonRoot["previous-postcode"];
```
_should_ have worked, but there seems to be a defect for the assignment overload such that this doesn't compile. For non const json values, the bracket notation actually returns a proxy, and the assignment operator on the proxy appears to have an issue with overloads. I need to look at that.

But you can do either of these:
```
newJsonDoc["previous-postcode"] = originalJsonDoc.at("previous-postcode");

newJsonDoc.try_emplace("previous-postcode", originalJsonDoc.at("previous-postcode"));
```
Both will give you
```
{"previous-postcode":{"postcode1":0,"postcode2":5,"postcode3":9}}
```
(The second is more efficient.)

**Reply by jasonmnemonic** [reply]
Thanks Daniel for the fast reply! :-)

Another question, will both the operations `.at` and `.try_emplace` cause a move operation? I mean I would like the keep the original json structure unchanged (not removed).

Ta.

**Reply by danielaparker** [reply]
No. For a move operation, you would have to explicitly write
```
newJsonDoc["previous-postcode"] = std::move(originalJsonDoc.at("previous-postcode"));

newJsonDoc.try_emplace("previous-postcode", std::move(originalJsonDoc.at("previous-postcode")));
```

**Reply by jasonmnemonic** [reply]
Thanks Dan. Sorry, I think I asked a wrong question. 

I mean if I want to avoid a move operation where I want to copy contents i.e. after doing `newJsonDoc.try_emplace("previous-postcode", originalJsonDoc.at("previous-postcode"));` I would like the `"previous-postcode"` node to still exist in `originalJsonDoc`.

I had a look at the code, I think `newJsonDoc["previous-postcode"] = originalJsonDoc.at("previous-postcode");` or `try-emplace` should be OK the `operator=` initialises from the `other`.

If I read it wrongly, please correct me. Thank you.

**Reply by danielaparker** [reply]
Regarding "I would like the "previous-postcode" node to still exist in originalJsonDoc", it will, 
```
newJsonDoc["previous-postcode"] = originalJsonDoc.at("previous-postcode");

newJsonDoc.try_emplace("previous-postcode", originalJsonDoc.at("previous-postcode"));
```
make copies, they do not move.
